### DISCUSSION TITLE
Unexplainable error when merging feature file containing reverse chaining single substitutions

### DISCUSSION THREAD

**Discussion by Panogen** [fontforge/fontforge#5186] [Q&A]
I am making a font for a fictional script which requires punctuation marks to be substituted for alternate forms when they appear next to alphanumeric characters. The feature file I use for this utilizes reverse chaining single substitutions in order to get these to work properly when in front of said characters:
```
lookup basic_punct_backwards {
	rsub @xymyric_all [period comma exclam question colon semicolon ellipsis]'
			by [xy_fullstop xy_commacn xy_exclamdown xy_questiondown xy_colon xy_semicolon xy_ellipsis];
} basic_punct_backwards;
```
FontForge has no issue with this lookup. However, I have a similar one right below this:
```
lookup quotes_backwards {
	rsub @xymyric_and_allnums [quotedbl quotedblleft]' by xy_kagileft;
} quotes_backwards;
```
and for some reason, FontForge doesn't like it. I get this error when merging the feature file:
![image](https://user-images.githubusercontent.com/73259774/214481433-a9d96be9-7424-4a64-a061-7b57b82538b1.png)
where line 54 refers to `rsub @xymyric_and_allnums...`.

I don't understand what the problem is. The format of both `rsub` statements is exactly the same, yet FontForge doesn't like one of them. I have only marked a single glyph class as the target of substitution in either example, so the error message makes no sense.

[Here is the feature file in question.](https://github.com/Panogen/sorazora_serif/blob/1346e96a94780f798db769e78f22ec6b0632df6f/sources/feature/punctuation.fea)

**Comment by ctrlcctrlv** [comment]
_As far as FontForge and the font format is concerned_, these statements are very different. One is a many-to-many and one is a many-to-one.

The first one means: for every character in the `@xymyric_all` class, substitute the matching class with one of the characters in the `[xy_fullstop xy_commacn xy_exclamdown xy_questiondown xy_colon xy_semicolon xy_ellipsis]` class (**many-to-many**).

The second one means: for every character in the `@xymyric_and_allnums` class, substitute the match with the single character `xy_kagileft` (**many-to-one**).

I'm not sure exactly what's wrong, but it's very important to note that totally different lookups need to be used for the two different cases. In OpenType, lookup types are numbered. And, remember that lookups can trigger other lookups. I assume it looks like in the font data:

```
GSUB Lookup Type Reverse Chaining Single Substitution (8)
Lookahead: @xymyric_all
Match: [period comma exclam question colon semicolon ellipsis]
Substitute: Trigger lookup type 1 (single substitution)

Lookup Type 1.2 (Single Substitution Format 2)
Match: [period comma exclam question colon semicolon ellipsis]
Substitute: [xy_fullstop xy_commacn xy_exclamdown xy_questiondown xy_colon xy_semicolon xy_ellipsis]
```

Meanwhile the second statement looks like this:

```
GSUB Lookup Type Reverse Chaining Single Substitution (8)
Lookahead: @xymyric_and_allnums
Match: [quotedbl quotedblleft]
Substitute: Trigger lookup type 2 (multiple substitution)

Lookup Type 2 (Multiple Substitution Format 1)
Match: [quotedbl quotedblleft]
Substitute: xy_kagileft
```

**Reply by ctrlcctrlv** [reply]
@Panogen Because in the standard there is only a `ReverseChainSingleSubstFormat1` other software might be simplifying the case into multiple single substitutions.

I'd perhaps try:

```fea
lookup quotes_backwards {
	rsub @xymyric_and_allnums quotedbl' by xy_kagileft;
	rsub @xymyric_and_allnums quotedblleft' by xy_kagileft;
} quotes_backwards;
```

Also consider avoiding `rsub` if unneeded. It's an advanced OTL lookup type and I'm pretty sure you can rewrite this as a regular `sub`, which is supported in more software.

**Reply by Panogen** [reply]
I don't now why I didn't think to do that—silly me. With this change the features merge no problem, but now FontForge crashes when I try to generate the font.

Also, all the lookups did just use `sub` originally. The reason `rsub` is used is I don't see another way to get this:
![image](https://user-images.githubusercontent.com/73259774/214487405-43ac9cb1-7a10-499f-999e-02fbf7fabf03.png)
to become this:
![image](https://user-images.githubusercontent.com/73259774/214487435-112389eb-3790-4634-b92f-7a584450adfd.png)
and not this:
![image](https://user-images.githubusercontent.com/73259774/214487469-a0da294a-a221-4cbf-9691-6ed4b57f4e4b.png)

The only way I can think of to make it work is to start at the letter sequence and work backwards through the punctuation characters.

**Reply by ctrlcctrlv** [reply]
Beautiful font. I am afraid you're in the processing of discovering what I did a few years ago, that FontForge is just not up to the task of modern complex OpenType Layout table generation, and saving your .fea files separately and merging them ex-post-facto with `fontmake` works far more reliably.

**Reply by Panogen** [reply]
Okay, I guess it's time to learn how `fontmake` works then. Thanks for the tips. I wonder if `fontmake` will let me just use the one-liner instead of having to break it up.

**Reply by ctrlcctrlv** [reply]
But as far as that query is concerned, there's also no shame in abusing multiple glyphs that look the same for OTL purposes, or blank separator glyphs.

Let's say you make a 0-width glyph with zero length called `null_mark`.

You can do this:

```fea
lookup pad_dot {
    sub \dot by \dot \null_mark;
} pad_dot;

lookup context_dot {
    sub \dot' lookup pad_dot [@xymyric_all];
} context_dot;
```

You can set the GDEF class of `null_mark` to be `mark`. Then if you 

```fea
lookup rep_dot {
    sub \dot by \hollow_dot;
} rep_dot;
lookup __0 {
    sub \dot' lookup rep_dot \null_mark;
} __0;
lookup __1 {
    sub \dot' lookup rep_dot \hollow_dot;
} __1;

feature ccmp {
    lookup context_dot;
    lookup __0;
    lookup __1;
} ccmp;
```

Or something similar you may be able to do it w/o `rsub`. This may seem ridiculous, if so sorry! But it can sometimes be worth it to avoid `rsub`.

**Reply by ctrlcctrlv** [reply]
Oh, another fun `.fea` keyword is:

```fea
subtable;
```

That splits the current lookup subtable in two and is sometimes mandatory for more obscure cases.

Happy hacking!~

**Comment by ctrlcctrlv** [comment]
Oh, one more tip: if you insist on doing such advanced OTL tricks (no shame in that, a lot of my fonts do too), you may want to consider offloading OpenType compilation from FontForge onto `fontmake`. `fontmake` is generally a lot more accurate and chooses better, smaller lookups in more cases. See also [simoncozens/fez](https://github.com/simoncozens/fez).
